src/snyk/analytics/itly.ts

```diff
@@ -0,0 +1,169 @@
+    this.anonymousId = uuidv4();
+  }
+
+  public load(): Iteratively | null {
+    if (!this.shouldReportEvents) {
+      return null;
+    }
+
+    let { segmentWriteKey } = require('../../../snyk.config.json') as { segmentWriteKey: string };
+    if (!segmentWriteKey) {
+      this.logger.debug('Segment analytics write key is empty. No analytics will be collected.');
+      return this;
+    } else if (this.isDevelopment) {
+      segmentWriteKey = process.env.SNYK_VSCE_SEGMENT_WRITE_KEY ?? '';
+    }
+
+    const segment = new SegmentPlugin(segmentWriteKey);
+    const isDevelopment = this.isDevelopment;
+
+    itly.load({
+      disabled: !this.shouldReportEvents,
+      environment: isDevelopment ? 'development' : 'production',
+      plugins: [segment, new ItlyErrorPlugin(this.logger)],
+    });
+
+    this.loaded = true;
+
+    return this;
+  }
+
+  public identify(): void {
+    if (!this.canReportEvents()) {
+      return;
+    }
+
+    itly.identify(this.anonymousId);
+  }
+
+  public alias(userId: string): void {
+    if (!this.canReportEvents()) {
+      return;
+    }
+
+    this.userId = userId;
+
+    const { version } = require('../../../package.json') as { version: string };
+
+    // Calling identify again is the preferred way to merge authenticated user with anonymous one,
+    // see https://snyk.slack.com/archives/C01U2SPRB3Q/p1624276750134700?thread_ts=1624030602.128900&cid=C01U2SPRB3Q
+    itly.identify(
+      this.userId,
+      {},
+      {
+        segment: {
+          options: {
+            anonymousId: this.anonymousId,
+            context: {
+              app: {
+                name: this.ide,
+                version,
+              },
+              testIde: this.ide, // first try
+            },
+            traits: {
+              testIde: this.ide, // first try
+            },
+            testUserIde: this.ide, // second try
+          },
+          testUserIdeRandom: this.ide, // third try
+        },
+      },
+    );
+  }
+
+  public issueIsViewed(properties: IssueIsViewedProperties): void {
+    if (!this.canReportEvents() || !this.userId) {
+      return;
+    }
+
+    itly.issueIsViewed(this.userId, properties);
+  }
+
+  public analysisIsReady(properties: AnalysisIsReadyProperties): void {
+    if (!this.canReportEvents() || !this.userId) {
+      return;
+    }
+
+    itly.analysisIsReady(this.userId, properties);
+  }
+
+  public analysisIsTriggered(properties: AnalysisIsTriggeredProperties): void {
+    if (!this.canReportEvents() || !this.userId) {
+      return;
+    }
+
+    itly.analysisIsTriggered(this.userId, properties);
+  }
+
+  public welcomeViewIsViewed(): void {
+    if (!this.canReportEvents()) {
+      return;
+    }
+
+    const userId = this.userId ?? this.anonymousId;
+
+    itly.welcomeIsViewed(userId, {
+      ide: this.ide,
+    });
+  }
+
+  public pluginIsInstalled(): boolean {
+    if (!this.canReportEvents()) {
+      return false;
```

Comment:
I might not understand the logic behind it, so sorry for stupid questions or remarks...
It sounds a bit strange that we presume !pluginIsInstalled when we !canReportEvents

Reply:
The class is responsible for reporting telemetry analytics. We don't want to report `pluginIsInstalled` event when we are not allowed/cannot report it due to disabled user setting ("snyk.yesTelemetry": false) or when the telemetry is not loaded (`this.loaded` prop).

What could help the code to be more readable? I didn't want to introduce a generic method to report all events because it would simply be a union of all Itly generated typings.

Reply:
See it now, so `pluginIsInstalled` is an Event. May be pre/post fixing all events with `E(e)vent` will help with readability? Not sure about TS  naming standards...
Anyway it's just minor thing, feel free to ignore it.

Reply:
@ArtsiomCh yes, it's an event. I've reused and kept the naming pattern guys have at `registry` for consistency reasons, but I agree that it'd be better if the method would reflect an action that it does, instead of having simply an event name in it.

I've added a `log` prefix rather than `event` to couple the action to the object.